### DISCUSSION TITLE
How do I make completions on <Tab> EXACTLY like Supertab?

### DISCUSSION THREAD

**Discussion by OracleMachine** [LazyVim/LazyVim#875] [Q&A] [answered]
The previous suggested [solution](https://github.com/LazyVim/LazyVim/discussions/39), works almost perfect but with only one catch. 

After apply the solution mentioned above, the first time you tap <TAB>, it didnt insert with the first item in completion list. It skip the first item and then insert the second item.  In other words, I have to press enter if I want to insert the first item.

However, In supertab, tap <TAB> always cycling from the every first item in the matchings.

**Comment by dpetka2001** [comment, marked as answer]
I answered to your question in the other thread. Please check to see if that solution works. If it doesn't, please wait for someone more experienced to suggest you a more appropriate one.

**Reply by OracleMachine** [reply]
Wonderful! I put the link to the "other thread" here in case people missed the context.

[Solution by @dpetka2001 ](https://github.com/LazyVim/LazyVim/discussions/39#discussioncomment-6069716)

Also, @jyuan0  deserves special thanks for the [diligent efforts in advancing the original solution](https://github.com/LazyVim/LazyVim/discussions/875#discussioncomment-6076237).

**Comment by jyuan0** [comment]
I was curious, so I took a look. The answer in the other thread works for me:
```
      opts.preselect = cmp.PreselectMode.None
      opts.completion = {
        completeopt = "noselect",
      }
```

The only change I made was to add the `completeopt` that LazyVim sets (`menu,menuone,noinsert`):
```
      opts.completion = {
        completeopt = "menu,menuone,noinsert,noselect",
      }
```

One thing I noticed with this change was that `Enter` still completes the first item, even though nothing is explicitly selected. That can be changed by modifying the `<CR>` mapping: https://github.com/hrsh7th/nvim-cmp/wiki/Example-mappings#safely-select-entries-with-cr

**Reply by dpetka2001** [reply]
The `opts` get merged with the parent `cmp` spec so you're not overwriting anything this way. You're only adding to `opts.completion.completeopt` the `noselect` value. The rest are still there (I think, pls correct me if I'm wrong). But it doesn't hurt that you add them again as well.  Nice trick for the `<CR>` mapping. Good to know :smile:

**Reply by jyuan0** [reply]
At first I thought you were right, but running `require'cmp.config'.get().completion.completeopt` just returns `noselect` (it doesn't seem to be merging).

Sorry, reposting here because it should have been a reply

**Reply by dpetka2001** [reply]
Thank you for clarifying this. I didn't know you could inspect the values this way. Out of curiosity, how did you find this? Is it somewhere in `nvim-cmp`'s docs?

**Reply by jyuan0** [reply]
Thank you for answering the original question!

I'm not a Lua or Neovim expert, so maybe I'm missing something or wrong about this, but I found this by searching the `nvim-cmp` repo for `completeopt` (on GitHub). One of the results was a commit which used that call ([1e7355a5fb5925dfd88bc60cefa3157169d09798](https://github.com/hrsh7th/nvim-cmp/commit/1e7355a5fb5925dfd88bc60cefa3157169d09798)).

**Reply by dpetka2001** [reply]
Thank you for that. I have one more question. There are some entries that have functions as values. Do you happen to know how we can read those function values? For examples in mappings it says
```lua
  mapping = {
    ["<C-B>"] = {
      i = <function 4>
    },
  }
```
Is there a way to see the contents of `<function 4>`?

**Reply by jyuan0** [reply]
If you want to see the function definition/source, I think you would need to drop into a debugger.

You could also try the debug and introspection libraries in the Lua standard libraries (https://www.lua.org/pil/23.1.html):
```
:lua =debug.getinfo(require'cmp.config'.get().mapping["<C-B>"].i)
```

It's not as good as a debugger, but the info it returns might be helpful enough.

**Reply by dpetka2001** [reply]
Thanks I will look into that :smile:

**Comment by theol0403** [comment]
I've made a PR with the snippets from this thread!

https://github.com/LazyVim/lazyvim.github.io/pull/113
